app/javascript/Pages/Classes.jsx

```diff
@@ -1,13 +1,206 @@
+    }
+    setSkillFilter("");
+  };
+
+  const fetchClasses = async () => {
+    const params = new URLSearchParams();
+
+    if (difficultyFilter) {
+      params.append("difficulty", difficultyFilter);
+    }
+    if (categoryFilter) {
+      params.append("category_id", categoryFilter);
+    }
+    if (skillFilter) {
+      params.append("skill_id", skillFilter);
+    }
+
+    let url = "/api/classes";
+
+    if (params.toString()) {
+      console.log(params.toString());
+      url += "?" + params.toString();
+    } else {
+      console.log("no parameters");
+    }
+
+    console.log(url);
+    let classes = [];
+    try {
+      const response = await fetch(url);
+      classes = await response.json();
+    } catch (error) {
+      console.log(error);
+      classes = [];
+    }
+
+    const components = classes.map((c) => {
+      return <p>{c.title}</p>;
+    });
+
+    setClasses(components);
+  };
+
+  const addCategoryButtons = async () => {
+    let categories = [];
+    try {
+      const response = await fetch(`/api/skills`);
+      categories = await response.json();
+    } catch (error) {
+      console.log(error);
+      categories = [];
+    }
+
+    const categoryComponents = [];
+
+    categories.map((c, i) => {
+      categoryComponents.push(
+        <ButtonFilter
+          props={{
+            onClick: updateCategory,
+            params: c,
+          }}
+        >
+          {c.name}
+        </ButtonFilter>
+      );
+    });
+
+    setCategories(categoryComponents);
+  };
+
+  const addDifficultyButtons = () => {
+    const buttons = [
+      <ButtonFilter
+        props={{
+          onClick: setDifficultyFilter,
+          params: "beginner",
+        }}
+      >
+        Beginner
+      </ButtonFilter>,
+      <ButtonFilter
+        props={{
+          onClick: setDifficultyFilter,
+          params: "intermediate",
+        }}
+      >
+        Intermediate
+      </ButtonFilter>,
+      <ButtonFilter
+        props={{
+          onClick: setDifficultyFilter,
+          params: "advanced",
+        }}
+      >
+        Advanced
+      </ButtonFilter>,
+    ];
+
+    return buttons;
+  };
+
+  const skillsPopup = (skills) => {
+    const components = skills.map((s) => (
+      <p
+        onClick={() => {
+          setSkillFilter(s.id);
+          setSkills([]);
+        }}
+      >
+        {s.name}
+      </p>
+    ));
+    components.unshift(
+      <p
+        onClick={() => {
+          setSkillFilter("");
+          setSkills([]);
+        }}
+      >
+        All skills
+      </p>
     );
-  }
-}
+    setSkills(components);
+  };
+
+  useEffect(() => {
+    addCategoryButtons();
+  }, []);
+
+  useEffect(() => {
+    fetchClasses();
+  }, [difficultyFilter, categoryFilter, skillFilter]);
+
+  return (
+    <Layout current_user={props.current_user}>
+      <div className="filter-wrapper">
+        <ButtonFilter
+          props={{
+            onClick: updateCategory,
+            params: "",
```

Comment:
**suggestion (non-blocking)**: in the ButtonFilter component you can define the defaultProps so that you don't need to pass empty strings here

Reply:
It can be this? 
`props.params ||= "";`

Reply:
yes you could do `props.params || ""` (without the = as you should not change the props), but the cleaner solution is to use defaultProps https://reactjs.org/docs/typechecking-with-proptypes.html#default-prop-values

Reply:
But in that case we should define default values for all props? Like onClick and id? What would we put there?

Reply:
I think you only need to define defaults for optional props, onClick should default to an emptu function `() => {}`

Reply:
But can we do something like
`Component.defaultProps = {
  props.params: ""
}`
? I think we tried it and didn't work
Otherwise we'd have to define a default value for the whole props object

Reply:
in this case it would be:
```
Classes.defaultProps = {
  params: ""
};
```